src/pam.rs

```diff
@@ -98,6 +106,47 @@ impl PamLibExt for Pam {
         let pointer = get_item(self.0, PamItemType::RUSER)?;
         unsafe { Ok(pointer.map(|p| CStr::from_ptr(p as *const c_char))) }
     }
+
+    fn conv(&self, prompt: Option<&str>, style: PamMsgStyle) -> PamResult<Option<&CStr>> {
+        let pointer = match get_item(self.0, PamItemType::CONV) {
+            Ok(Some(p)) => p,
+            Ok(None) => return Ok(None),
+            Err(e) => return Err(e),
+        };
+        let conv = unsafe { &*(pointer as *const PamConv) };
+
+        let mut resp_ptr: *mut PamResponse = ptr::null_mut();
+        let msg_cstr = CString::new(prompt.unwrap_or("")).unwrap();
+        let msg = PamMessage {
+            msg_style: style,
+            msg: msg_cstr.as_ptr(),
+        };
+
+        match conv
+            .cb
+            .map(|cb| {
+                cb(
+                    1,
+                    &mut (&msg as *const PamMessage),
+                    &mut resp_ptr,
+                    conv.appdata_ptr,
+                )
+            })
+            .map(|ret| PamError::new(ret))
+        {
+            Some(PamError::SUCCESS) => {
+                // PamResponse.resp is null for styles that don't return user input like PAM_TEXT_INFO
```

Comment:
If that's the case, the ```resp``` member should be an ```Option<*mut c_char>``` in ```struct PamResponse```. As a bonus you can then replace all the ```if/else``` below with a single ```response.map()``` since rust will turn the null pointer into a ```None```

Reply:
Should we do the same for `PamMessage.msg`, assuming upstream allows it:
```rust
#[repr(C)]
pub struct PamMessage {
    pub msg_style: PamMsgStyle,
    pub msg: Option<*const c_char>,
}


```

Reply:
This also did not play nice with the `impl PamResponse` in `pam_types.rs`, so I purged it. If this is an important part of the C ABI, I can fixup the logic.

Reply:
I can't even remember why I wrote this ```impl PamResponse``` since ```pam_conv``` was not implemented.
This was supposed to allow for cleaning up secrets from memory after they're used. But this should probably be the responsibility of the client anyway since it's the one allocating and freeing the responses. You were right to get rid of it.